### PR TITLE
First round of gRPC api v2 functions

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Implement some of the new gRPC api v2 functions. https://github.com/Concordium/concordium-node/issues/433
Should be merged with https://github.com/Concordium/concordium-node/pull/482

## Changes

- Extend the gRPC api v2 service with `GetModuleList`, `GetModuleSource`, `GetInstanceList`, `GetInstanceInfo`, `GetNextAccountSequenceNumber`, `GetConsensusInfo`, `GetAncestors` and some initial work for `GetTransactionStatus`.
- Added the relevant types.
- Introduce `AccountIdentifierInput` and updated `AccountInfoRequest` to use it.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by abizjak — COMMENTED** [review]
Looks good overall, but it seems to me that the transaction status related stuff, that is not complete, should not be part of this?

File: v2/concordium/service.proto [hunk 1]
```diff
@@ -21,4 +21,35 @@ service Queries {
 
   // Retrieve the list of accounts that exist at the end of the given block.
   rpc GetAccountList (BlockHashInput) returns (stream AccountAddress);
+
+  // Get a list of all smart contract modules. The stream will end
+  // when all modules that exist in the state at the end of the given
+  // block have been returned.
+  rpc GetModuleList (BlockHashInput) returns (stream ModuleRef);
+
+  // Get ancestors for the provided block.
```
Comment (Bargsteen) [inline v2/concordium/service.proto:30]:
```suggestion
  // Get ancestors for the provided block.
  // Get a list of ancestors of the provided block. The stream will end
  // when the requested number of ancestors has been returned. Or, alternatively, when all ancestors have been  
  // returned if there are fewer total ancestors than requested.
```

File: v2/concordium/types.proto [hunk 2]
```diff
@@ -462,6 +492,300 @@ message FinalizedBlockInfo {
...
+  message Finalized {
+    TransactionSummaryInBlock outcomes = 1; // TODO: Should this also be repeated for when finalization committee is corrupt?
+  }
+
+  oneof status {
+    Empty received = 1;
+    Committed committed = 2;
+    Finalized finalized = 3;
+  }
+}
```
[the remark this answers is not in the document]
Comment (Bargsteen) [reply v2/concordium/types.proto:600]: These have been resolved in my branch. We talked about removing the transaction status parts of this PR, but that will likely lead to more merge conflicts for my PR. But if you deem it necessary to remove it, we can do that.

File: v2/concordium/types.proto [hunk 3]
```diff
@@ -462,6 +492,300 @@ message FinalizedBlockInfo {
...
+  BlockHash genesis_block = 2;
+  // Time of the (original) genesis block.
+  Timestamp genesis_time = 3;
+  // (Current) slot duration in milliseconds.
+  Duration slot_duration = 4;
+  // (Current) epoch duration in milliseconds.
+  Duration epoch_duration = 5;
+  // Hash of the last finalized block.
+  BlockHash last_finalized_block = 6;
+  // Absolute height of the best block.
```
Comment (Bargsteen) [inline v2/concordium/types.proto:735]: Consider mentioning that this means from the original genesis block. Also for the field below.

File: v2/concordium/types.proto [hunk 4]
```diff
@@ -462,6 +492,300 @@ message FinalizedBlockInfo {
...
+  optional Timestamp block_last_arrived_time = 16;
+  // Moving average latency between a block's slot time and its arrival.
+  double block_arrive_latency_ema = 17;
+  // Standard deviation of moving average latency between a block's slot time and its arrival.
+  double block_arrive_latency_emsd = 18;
+  // Moving average time between block arrivals.
+  optional double block_arrive_period_ema = 19;
+  // Standard deviation of moving average time between block arrivals.
+  optional double block_arrive_period_emsd = 20;
+  // Moving average number of transactions per block.
```
Comment (Bargsteen) [inline v2/concordium/types.proto:763]:
```suggestion
  // Exponential moving average number of transactions per block.
```
Also for the other ones. That will explain the abbreviations (ema/emsd).

File: v2/concordium/types.proto [hunk 5]
```diff
@@ -462,6 +492,300 @@ message FinalizedBlockInfo {
...
+  uint32 finalization_count = 23;
+  // Time of last verified finalization.
+  optional Timestamp last_finalized_time = 24;
+  // Moving average time between finalizations.
+  optional double finalization_period_ema = 25;
+  // Standard deviation of moving average time between finalizations.
+  optional double finalization_period_emsd = 26;
+  // Currently active protocol version.
+  ProtocolVersion protocol_version = 27;
+  // The number of chain restarts via a protocol update. An effected
```
Comment (Bargsteen) [inline v2/concordium/types.proto:777]:
```suggestion
  // The number of chain restarts via a protocol update. A completed
```

File: v2/concordium/types.proto [hunk 6]
```diff
@@ -462,6 +492,300 @@ message FinalizedBlockInfo {
...
+  // protocol update instruction might not change the protocol version
+  // specified in the previous field, but it always increments the genesis
+  // index.
+  GenesisIndex genesis_index = 28;
+  // Block hash of the genesis block of current era, i.e., since the last protocol update.
+  // Initially this is equal to 'csGenesisBlock'.
```
Comment (Bargsteen) [inline v2/concordium/types.proto:783]:
```suggestion
  // Initially this is equal to 'genesis_block'.
```

File: v2/concordium/types.proto (hunk 3 above)
Comment (limemloh) [reply v2/concordium/types.proto:735]: I find it a bit noisy to explain the definition of absolute height everywhere we use it and I would prefer to only have it next to the type definition.

File: v2/concordium/types.proto [hunk 7]
```diff
@@ -462,6 +492,300 @@ message FinalizedBlockInfo {
...
+message InitName {
+  string value = 1;
+}
+
+// A smart contract v0 state.
+message ContractState {
+  bytes value = 1;
+}
+
+message TransactionStatus {
```
[the remark this answers is not in the document]
Comment (limemloh) [reply v2/concordium/types.proto:585]: This has been removed

File: v2/concordium/service.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (limemloh) [reply v2/concordium/service.proto:30]: I have added to this now

File: v2/concordium/types.proto [hunk 8]
```diff
@@ -40,6 +40,16 @@ message AccountIndex {
   uint64 value = 1;
 }
 
+// A smart contract module reference. This is always 32 bytes long.
+message ModuleRef {
+  bytes value = 1;
+}
+
+// Source bytes of a smart contract module.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply v2/concordium/types.proto]: I have made a type, capturing the module version and the bytes of the wasm module

**Review by abizjak — APPROVED** [review]

**Pull request merged**